packages/gatsby/src/commands/serve.js

```diff
@@ -40,6 +41,29 @@ const readMatchPaths = async program => {
   return JSON.parse(rawJSON)
 }
 
+/**
+ * Adds trailing slash to request.url, in order to deceive static middleware
+ * into getting a directory request, thus preventing redirect.
+ * This is an express middleware to be used before static middleware
+ * @param {*} req
+ * @param {*} res
+ * @param {*} next
+ */
+const directoryRedirectPrevention = (req, res, next) => {
+  // url is home page
+  if (req.url === `/`) return next()
+  // url has trailing slash, no need to intervene
+  if (req.url.slice(-1) === `/`) return next()
+  // url is a file request, no need for trailing slash
+  if (fileExtensionRegEx.test(req.url)) return next()
+
+  // turning /slug into /slug/
+  req.originalUrl += `/` // https://expressjs.com/en/api.html#req.originalUrl
+  req.url += `/`
```

Comment:
Which made me look into `.url`, so. Same link has

> req.url is not a native Express property, it is inherited from Node’s http module.

> https://nodejs.org/api/http.html#http_message_url
> > When request.url is '/status?name=ryan'

So it looks like `.url` has the same problem. Is there a reason why this is not a problem?